Add 'low' & 'high' options for the gpio direction setting

macguru2000 opened the pull request:
This pull request is a proposed fix for issue #25 by allowing the developer to choose which state the gpio pin should be initially set to. This is especially helpful for dealing with certain hardware setups that operate in reverse logic (0 is hot, 1 is cold).

Thanks!

macguru2000 commented on rpi-gpio.js:
```diff
@@ -90,6 +90,8 @@ function Gpio() {
 
     this.DIR_IN   = 'in';
     this.DIR_OUT  = 'out';
+    this.DIR_LOW  = 'low';
```
A value for gpio which will set the direction to out and the initial value to 0

macguru2000 commented on rpi-gpio.js:
```diff
@@ -90,6 +90,8 @@ function Gpio() {
     this.DIR_IN   = 'in';
     this.DIR_OUT  = 'out';
+    this.DIR_LOW  = 'low';
+    this.DIR_HIGH = 'high';
```
A value for gpio which will set the direction to out and the initial value to 1

macguru2000 commented on rpi-gpio.js:
```diff
@@ -143,7 +145,7 @@ function Gpio() {
         }
 
-        if (direction !== this.DIR_IN && direction !== this.DIR_OUT) {
+        if (direction !== this.DIR_IN && direction !== this.DIR_OUT && direction !== this.DIR_LOW && direction !== this.DIR_HIGH) {
```
Don't throw an error when using either 'low' or 'high' values, which are perfectly acceptable values

unimonkiez commented:
???

JamesBarwell commented:
Thanks for going to the effort of submitting this patch and I'm sorry it has taken so long to get back to you.

I'm keen to merge, but is it possible please to add some extra test coverage just so that we can see the module working with these new options? There are some existing tests for direction and I think they can just be copied.

Also is there any chance you could update the readme to provide some info for the extra options?

Cheers.

karlstulik commented:
I found that channel 7 is the only one that defaults to high, the rest (11, 12, 13, 15, 16, 18, 22) defaulted to low.  I also found that the proposed fix does not work for me.
I used it like this:
`gpio.setup(11, gpio.DIR_LOW, gpio.EDGE_BOTH);`

macguru2000 commented:
Hey guys, sorry I haven't been watching this thread lately.

Anyways, I'm happy to add some test coverage and explain in the read me what I am trying to achieve with this pull request.

More coming soon...

macguru2000 commented:
@karlstulik, I did not notice the EDGE_ options when originally using this fix. My code skips the edge option altogether and just has a callback function. Perhaps not using a callback function is contributing to it not working for you? Mine works like this:

```
function startServer() {
  /*
  code that starts an http server and listens for requests to turn 
  the pin on and off. the pin start "hot" because I used HIGH as the
  direction.
  */ 
}

gpio.setup(7, gpio.DIR_HIGH, startServer);
```

I don't have a good setup to test right at the moment, but if memory serves me correctly, using DIR_HIGH was necessary for me, and made the pin start "hot", which I remember it wasn't doing originally.

macguru2000 commented on README.md:
```diff
@@ -34,7 +34,7 @@ Please note that there are two different and confusing ways to reference a chann
 Sets up a channel for read or write. Must be done before the channel can be used.
 * channel: Reference to the pin in the current mode's schema.
-* direction: The pin direction, pass either DIR_IN for read mode or DIR_OUT for write mode. Defaults to DIR_OUT.
+* direction: The pin direction, pass either DIR_IN for read mode or DIR_OUT for write mode. You can also pass DIR_LOW or DIR_HIGH to use write mode with a specific initial "hot" state. Defaults to DIR_OUT.
```
Add DIR_LOW & DIR_HIGH to the API Docs

macguru2000 commented on README.md:
````diff
@@ -103,6 +103,23 @@ function write() {
 }
 ```
 
+### Setup and write to a pin that starts as on
````
Give an example that uses the DIR_HIGH setting for direction. Didn't bother making a DIR_LOW one because it would be so similar.

macguru2000 commented on test/main.js:
```diff
@@ -241,6 +241,30 @@ describe('rpi-gpio', function() {
                 });
             });
 
+            context('and direction is specified low-ward', function() {
```
tests that DIR_HIGH and DIR_LOW do indeed set the pin's direction to 'high' and 'low' respectively.

macguru2000 commented:
@JamesBarwell, I've added the requested tests and added some documentation to the read me. the tests passed on my Pi running NodeJS 4.4.0. I did not run or edit the integration test since the Pi is not wired properly for it.

Let me know if you need me to do anything else. Thanks again.
Cheers! 🍺 

Zach

JamesBarwell commented:
Apologies for the delay, but I wanted to give it a proper test myself. I've been testing the patch this morning on my Model B running node 6.2.1 and I'm happy with it. I'll update in a bit with new module version details. Much appreciated, cheers.

JamesBarwell commented:
Published as `rpi-gpio@0.8.0`

unimonkiez commented:
Tested this in my [library](https://github.com/unimonkiez/desktop-controller) on raspberry pi 1 model b, works great! Huge thanks
